File: 202-happy-number/happy-number.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function (n) {
    let slow = n;
    let fast = n;

    do {
        slow = square(slow);
        fast = square(square(fast));
    } while (slow !== fast);

    return slow === 1;
};

function square(n) {
    let sum = 0;
    while (n > 0) {
        let digit = n % 10;
        sum += (digit * digit);
        n = Math.floor(n / 10);
    }

    return sum;
}

File: 202-happy-number/happy-number.java
class Solution {
    public boolean isHappy(int n) {
        int slow = n;
        int fast = n;

        do {
            slow = square(slow);
            fast = square(square(fast));
        }while(slow != fast);

        return slow == 1;
    }

    public int square(int n){
        int sum = 0;
        while(n>0){
            int dig = n % 10;
            sum += dig * dig;
            n = n/10;
        }
        return sum;
    }
}